### ISSUE TITLE
V8: MemberTypeService.Save throws exception

### ISSUE THREAD

**Issue by mclausen** [umbraco/Umbraco-CMS#4963]
Hi Core team
I was tasked to solve [Umbraco-Deploy](https://github.com/umbraco/Umbraco-Deploy/issues/355),
which was fixed in [PR](https://github.com/umbraco/Umbraco-Deploy/pull/357). 

It revealed a potential error in core cms when saving a MemberType.

Reproduction
------------
To reproduce this bug you need to work from deploy, branch 'tasks/355-Variations-are-not-supported-on-members'

- Create a fresh v8 Cloud site
- Go into Kudu ~/wwwroot/data and run echo > deploy-export
- Clone down the project
- When spinning it up and restoring you retrieve the following error
  
```
Message: An item with the same key has already been added.

   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Umbraco.Core.Persistence.Factories.MemberTypeReadOnlyFactory.GetPropertyTypes(MemberTypeReadOnlyDto dto, MemberType memberType, Dictionary`2 standardProps)
   at Umbraco.Core.Persistence.Factories.MemberTypeReadOnlyFactory.BuildEntity(MemberTypeReadOnlyDto dto, Boolean& needsSaving)
   at Umbraco.Core.Persistence.Repositories.Implement.MemberTypeRepository.<BuildFromDtos>b__22_0(MemberTypeReadOnlyDto x)
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Umbraco.Core.Cache.FullDataSetRepositoryCachePolicy`2.GetAllCached(Func`2 performGetAll)
   at Umbraco.Core.Cache.FullDataSetRepositoryCachePolicy`2.GetAll(TId[] ids, Func`2 performGetAll)
   at Umbraco.Core.Services.Implement.ContentTypeServiceBase`3.GetAll(Int32[] ids)
   at Umbraco.Core.Services.Implement.ContentTypeServiceBase`3.ComposeContentTypeChanges(TItem[] contentTypes)
   at Umbraco.Core.Services.Implement.ContentTypeServiceBase`3.Save(TItem item, Int32 userId)
   at Umbraco.Deploy.Connectors.ServiceConnectors.ContentTypeConnector.MemberTypeConnector.SaveContentType(IMemberType contentType) in C:\Development\Umbraco\Umbraco-Deploy\Umbraco.Deploy\Connectors\ServiceConnectors\ContentTypeConnector\MemberTypeConnector.cs:line 118
```

**Comment by Shazwazza** [comment]
## Notes

Trying to make sense out of this has been 'interesting' but i think I've figure out the gist of it all.

### Background

When a member type is saved, we ensure that special properties are added - this is an absolute requirement. These properties are there to support both ASP.Net Membership Providers and ASP.Net Identity for members.

For some reason, our initial data install of Umbraco doesn't install 2 of these required properties: `umbracoMemberPasswordRetrievalAnswer`, `umbracoMemberPasswordRetrievalQuestion` ... **but it should**.

These property types are referenced a number of times in the Core code by their explicit property type aliases. If these aliases are missing, then they will be automatically re-added.

### Core issue

We are unfortunately performing a DB update when we are trying to do a DB read. We've had this problem in the past for some other data structure and it caused all sorts of problems and I believe this is also causing problems. What ends up happening is that we call `MemberTypeRepository.GetAll()`, this fetches some data, then checks against the resulting member types to see if they have all of the built in property types, if they don't  (which would happen on the first request to this since we don't install all required prop types by default [see above]), the code adds these missing property types and then calls `PersistUpdatedEntity` - which is now performing a DB write when all we were expecting was a DB read.

This problem goes a bit deeper too because when a call is made to simply save a member type, during this operation a call is made to GetAll(). So consider this:

* Member Type `A` is created (`MemberService.Save`)
* During creation, all required property types are added
* Member Type `A` values are changed: all property type aliases are changed to something random
* Member Type `A` is updated (`MemberService.Save`)
* During the update (`PersistUpdatedEntity`) a call is made to `PersistUpdatedBaseContentType` which calls GetAll()
   * During the GetAll() call, it builds the `MemberType`'s from the data in the database and notices that the required property types are missing because it can't find them by their required alias
   * The required property types are added to the model
   * This member type is then updated (`PersistUpdatedEntity`)
   * **Wat**? ... Yes, this means that we are now in a nested `PersistUpdatedEntity` operation for the same entity
   * Now that the inner update of this entity is done, it now contains duplicate built in properties, except the original ones are now just with some random alias
* Then the outer update of this entity continues and the result is cached

At some stage in this process, the internal dictionary `MemberType.MemberTypePropertyTypes` (which is used to track which property types are 'editable', 'visible' or 'sensitive') contains all of the original built in property aliases and then during one of the operations above when re-adding the built in properties because the original aliases were changed, we get the exception listed in this issue. Potentially part of this issue here is that this dictionary isn't considered when deep cloning for cache? Also it seems to me that `MemberTypePropertyProfileAccess` should be a struct?

### Deploy issue

Deploy has a process of renaming aliases like "myAlias" to "myAlias__0001". It does this for all aliases, whether it is at the content type level or the property type level and this is a problem for member types because of the special member type property types that must always exist and must always have the exact alias. Changing these aliases and persisting the entity results in the problem above.

### What's next?

I will try to write a unit test in the Core to replicate the issue and then we can figure out what to do from there.

**Comment by Shazwazza** [comment]
I have added the unit test to core to show the problem https://github.com/umbraco/Umbraco-CMS/commit/02ddf80014093fa4dc4b60c8dd01f954519c9b61

So the core will need to be fixed to deal with this correctly which I believe is 2 fold:

* I think we are not handling the `MemberType.MemberTypePropertyTypes` correctly
* We need to stop writing to the DB during a read operation

As for Deploy, we will need to discuss that since it should not be changing any of these built in property aliases.

**Comment by Shazwazza** [comment]
I have pushed more changes:

* Fixed issues with INotifyCollectionChanged collections which were wrong - and fixing this seems to have fixed issues above with the unit test
* Added code to enforce unique property type aliases - but have commented it out for now since if we put this in a patch release it is too risky
* Makes sure we don't write during a read, this is a big cause of some of these issues
* Property adds the system level property types to the outgoing member type with the correct property group

With the above, the exception is no longer thrown with deploy, however we need to decide how to handle member type property types with Deploy separately.

**Comment by zpqrtbnk** [comment]
Reviewing now. Which means, I had to merge with the changes that were made to content/media/member type repositories in `v8/dev`. Which means... I am trying to make sense of it all as *maybe* some of these changes broke a few things, too.

So, here is what I have now.

* Instanciate a new member type, it does *not* have a builtin properties
* Save a *new* member type, the missing builtin properties are added to that member type
* Save an *updated* member type, builtin properties are not handled (ie can be deleted)
* Fetch a member type, builtin properties that would be missing in db, are added (but not saved!)

I have updated and enabled the test that was ignored, and it *seems* to make sense now. As for Deploy... it is ok to rename the builtin properties from `alias` to `alias_0000`, save the member type, then rename back to `alias` and save again - as long as this is all on the *same* instance and Deploy does not *reload* the member type. Which, IIRC, is how we designed Deploy (the modified entities are kept and updated, never reloaded).

Going to push changes + do more tests.
